### PR TITLE
Restructure project and add Evaluation, Sagemaker Dockerfile, ++

### PR DESCRIPTION
**Pull request by finnss**
This pull request merges the changes done in my fork here: https://github.com/finnss/deep_learning_collection

Notable changes include:
- Renaming `hymenoptera` to `salmon_trout` universally, including the folder name
- Moving the Pipfile into the `salmon_trout` project folder (different projects likely want different pipfiles)
- Groups files into folders (azure, aws, evaluation, interpretability) for a more comprehensive structure
- Addition of the `evaluation` folder, containing code to run a model on validation data, as well as the evaluation code leading to graphs, statistics and matrixes for analysis
- A Dockerfile for use anywhere
- A Dockerfile and corresponding Jupyter Notebooks for training a model in a Docker container on Sagemaker
- A more useful README

The main `train.py` file is largely unchanged – the differences are mostly formatting automatically done by my IDE. The real changes are as follows:
- Adding a `shuffle` boolean parameter to the `load_datasets` function, which I needed for the evaluation script
- Making `final_layer` the default `--action`, which I think makes more sense
- Setting the default number of epochs to 10

*Known issues*
- The AWS Sagemaker scripts are currently broken, failing to find the correct dataset location.
- Moving the Azure code into its own folder might have broken imports or references; I have not tested this yet.
I suggest we tackle these issues in a separate PR.

If you have any questions, comments, our changes you disagree with, feel free to shoot!

### CODE REVIEW

File: salmon_trout/azure/run_azure_train.py [hunk 1]
```diff
@@ -20,7 +20,8 @@ def get_compute_target(workspace, priority):
         compute_config = AmlCompute.provisioning_configuration(
             vm_size="STANDARD_NC6", max_nodes=1, vm_priority=priority
         )
-        compute_target = ComputeTarget.create(workspace, CLUSTER_NAME, compute_config)
+        compute_target = ComputeTarget.create(
+            workspace, CLUSTER_NAME, compute_config)
```
Comment (khellan) [inline salmon_trout/azure/run_azure_train.py:24]: Why reformat this line, it's only 79 characters?

File: salmon_trout/evaluation/run_model.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    model.eval()
+    return model
+
+
+def download_image(url):
+    local_path = './test_images/%s' % url.split("/")[-1]
+    # local_path = './test_images/img.jpg'
+    if not os.path.isfile(local_path):
+        resp = requests.get(url, headers={
+                            'User-Agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/50.0.2661.102 Safari/537.36'})
```
Comment (khellan) [inline salmon_trout/evaluation/run_model.py]: Sneaky, Did you get blocked from some URLs without this user agent?

File: salmon_trout/evaluation/evaluate_model.py [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+        incorrect_guesses) - was_salmon_guessed_trout
+    # was_trout_guessed_salmon = len(
+    #     [guess for guess in incorrect_guesses if guess['guessed'] == 'salmon'])
+
+    confusion_matrix['tp'] = total_salmon_pics - was_salmon_guessed_trout
+    confusion_matrix['fp'] = was_trout_guessed_salmon
+    confusion_matrix['fn'] = was_salmon_guessed_trout
+    confusion_matrix['tn'] = total_trout_pics - was_trout_guessed_salmon
+
+    return confusion_matrix
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: Why not use sklearn.metrics.confusion_matrix?
https://scikit-learn.org/stable/modules/generated/sklearn.metrics.confusion_matrix.html#sklearn.metrics.confusion_matrix

File: salmon_trout/evaluation/evaluate_model.py [hunk 4]
```diff
@@ -0,0 +1,205 @@
...
+            #     all_guesses += [max(list(o.numpy()))]
+            total += labels.size(0)
+            correct += (predicted == labels).sum().item()
+
+            for j in range(inputs.size()[0]):
+                if labels[j] != predicted[j]:
+                    absolute_index = i * 4 + j
+                    filename, _ = dataloaders['val'].dataset.samples[absolute_index]
+                    incorrect_guess = {'index': absolute_index,
+                                       'guessed': class_names[predicted[j]], 'was': class_names[labels[j]], 'filename': filename}
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: Put each key, value pair on separate lines for readability

File: salmon_trout/evaluation/evaluate_model.py [hunk 5]
```diff
@@ -0,0 +1,205 @@
...
+
+    actual_labels = dataloaders['val'].dataset.targets
+
+    save_path = './graphs'
+    if not (os.path.exists(save_path) and os.path.isdir(save_path)):
+        os.system('mkdir %s' % save_path)
+    # now = datetime.now()
+    # dt_string = now.strftime("%Y-%m-%d_%H-%M-%S")
+    # save_path = 'graphs/%s/' % dt_string
+    # os.system('mkdir %s' % save_path)
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: Either remove commented code or explain why it's there.

File: salmon_trout/evaluation/evaluate_model.py [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+        [guess for guess in incorrect_guesses if guess['guessed'] == 'trout'])
+    was_trout_guessed_salmon = len(
+        incorrect_guesses) - was_salmon_guessed_trout
+    # was_trout_guessed_salmon = len(
+    #     [guess for guess in incorrect_guesses if guess['guessed'] == 'salmon'])
+
+    confusion_matrix['tp'] = total_salmon_pics - was_salmon_guessed_trout
+    confusion_matrix['fp'] = was_trout_guessed_salmon
+    confusion_matrix['fn'] = was_salmon_guessed_trout
+    confusion_matrix['tn'] = total_trout_pics - was_trout_guessed_salmon
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: If possible, use sklearn.metrics.confusion_matrix instead of writing own code.

File: salmon_trout/train.py [hunk 7]
```diff
@@ -225,28 +229,28 @@ def input_fn(request_body, content_type):
 if __name__ == "__main__":
     parser = argparse.ArgumentParser()
 
-    parser.add_argument("--epochs", type=int, default=1)
+    parser.add_argument("--epochs", type=int, default=10)
```
Comment (khellan) [inline salmon_trout/train.py]: The advantage of using 1 as default epochs is to ensure that you don't start a longish job when you just want to test this.

**Review by khellan — APPROVED** [review]
Just some minor comments. Review them and update if you agree. Then you are free to merge.

File: salmon_trout/azure/run_azure_train.py (hunk 1 above)
Comment (finnss) [reply salmon_trout/azure/run_azure_train.py:24]: An auto-formatter I have installed did this. Can disable if you prefer the single line

File: salmon_trout/evaluation/run_model.py (hunk 2 above)
Comment (finnss) [reply salmon_trout/evaluation/run_model.py]: heh yeah, pretty directly copy-pasted from a stack overflow thread. Maybe this shouldn't be in the repo

File: salmon_trout/evaluation/evaluate_model.py (hunk 3 above)
Comment (finnss) [reply salmon_trout/evaluation/evaluate_model.py]: I tried that initially, but I struggled to format my input in the way sklearn expected. I later had to conform anyway for the ROC-curve though so I can go back and check if that'll work.

File: salmon_trout/azure/run_azure_train.py (hunk 1 above)
Comment (khellan) [reply salmon_trout/azure/run_azure_train.py:24]: Odd, my VSCode says 79. Anyway, no big deal. Keep the line split

File: salmon_trout/evaluation/evaluate_model.py (hunk 3 above)
Comment (finnss) [reply salmon_trout/evaluation/evaluate_model.py]: Using the `y` and `y_pred` I had to generate for the ROC-curve worked nicely with sklearn's other functions. I replaced both the confusion matrix and recall/precision/f1-score generation with library methods now, thanks for the suggestion.

File: salmon_trout/evaluation/evaluate_model.py [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+    all_guesses = []
+    incorrect_guesses = []
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    with torch.no_grad():
+        print('Beginning evauluation using %s validation samples.' %
+              len(dataloaders["val"].dataset.targets))
+        for i, (inputs, labels) in enumerate(dataloaders["val"]):
+            if i % 10 == 0:
+                print('Sample %s...' % (i * 4))
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: This is a general comment. Use f-strings for variable substitutions in strings.

f'Sample {i * 4}' is much more readable than 'Sample %s...' % (i * 4)

File: salmon_trout/evaluation/evaluate_model.py [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+                    filename, _ = dataloaders['val'].dataset.samples[absolute_index]
+                    incorrect_guess = {'index': absolute_index,
+                                       'guessed': class_names[predicted[j]],
+                                       'was': class_names[labels[j]],
+                                       'filename': filename}
+                    incorrect_guesses += [incorrect_guess]
+
+    accuracy = correct / total
+    print('\nIncorrect guesses: \n', json.dumps(incorrect_guesses, indent=2))
+    print('Accuracy of the network on the %s validation images: %s' % (
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: Use f-string

File: salmon_trout/evaluation/evaluate_model.py [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+        total, accuracy))
+
+    dirpath = 'error_images'
+    if os.path.exists(dirpath) and os.path.isdir(dirpath):
+        shutil.rmtree('./%s' % dirpath)
```
Comment (khellan) [inline salmon_trout/evaluation/evaluate_model.py]: f-string

File: salmon_trout/evaluation/run_model.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+def load_model(model_path):
+    model = models.resnet18(pretrained=False, num_classes=2)
+    model.load_state_dict(torch.load(model_path))
+    model.eval()
+    return model
+
+
+def download_image(url):
+    local_path = './test_images/%s' % url.split("/")[-1]
```
Comment (khellan) [inline salmon_trout/evaluation/run_model.py]: Use f-string for variable substitutions in strings.

File: salmon_trout/train.py [hunk 12]
```diff
@@ -64,7 +64,8 @@ def train_model(
             epoch_loss = running_loss / dataset_sizes[phase]
             epoch_acc = running_corrects.double() / dataset_sizes[phase]
 
-            print("{} Loss: {:.4f} Acc: {:.4f}".format(phase, epoch_loss, epoch_acc))
+            print("{} Loss: {:.4f} Acc: {:.4f}".format(
```
Comment (khellan) [inline salmon_trout/train.py]: Use f-string

**Review by khellan — COMMENTED** [review]
Mostly looks good. Switch to f-strings for variable substitutions in string to get an approval.

**Review by khellan — APPROVED** [review]
